Implement a Ed25519 key pair check during the JWK import

javifernandez opened the pull request:
#### 6940fc8a24d037525d6a83c817016ad79489e2a2
<pre>
Implement a Ed25519 key pair check during the JWK import
<a href="https://bugs.webkit.org/show_bug.cgi?id=261323">https://bugs.webkit.org/show_bug.cgi?id=261323</a>

Reviewed by Žan Doberšek.

This change implemenets the logic to ensure they private and public
keys imported using JWK format are actually compatible, according to
the RFC8032 specification.

There are similar checks for the X25519, already implemented in both
the Mac and Gtk+ ports. The implementation of this check is also present
in the Mac port.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any.worker-expected.txt:
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
* Source/WebCore/crypto/gcrypt/GCryptRFC7748.cpp:
* Source/WebCore/crypto/gcrypt/GCryptRFC8032.cpp: Added.
(WebCore::validateEd25519KeyPair):
* Source/WebCore/crypto/gcrypt/GCryptRFC8032.h: Added.
* Source/WebCore/platform/SourcesGCrypt.txt:

Canonical link: <a href="https://commits.webkit.org/267957@main">https://commits.webkit.org/267957@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73b45711ff118453cd8c1492b4853a3ad798b739

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18489 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19058 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19996 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16996 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18953 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18610 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15807 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20874 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15834 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16560 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23074 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16853 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16731 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20957 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17303 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14655 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16392 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4328 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20755 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17150 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -151,17 +152,22 @@ std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmI
-    // FIXME: Implement this check for Ed25519
-
+    // No check
     return true;
```
By default, this should return false. That way the supported curves perform their own checks, and anything else would default to a failed check, regardless of the debug-specific assert.

javifernandez replied:
Done

zdobersek commented on Source/WebCore/crypto/gcrypt/GCryptRFC8032.cpp:
```diff
@@ -0,0 +1,105 @@
+            result.resize(32);
+        while (result.size() < 32)
+            result.append(0x00);
+        assert(result.size() == 32);
```
This should be ASSERT.

zdobersek commented on Source/WebCore/crypto/gcrypt/GCryptRFC8032.cpp:
```diff
@@ -0,0 +1,105 @@
+bool validateEd25519KeyPair(const Vector<uint8_t>& privateKey, const Vector<uint8_t>& publicKey)
+{
+    ASSERT(privateKey.size() == 32);
+    ASSERT(publicKey.size() == 32);
```
IMO, similar to the X25519 function, the sizes should be checked and false returned in either isn't 32.

javifernandez replied:
I had this check implemented in the 'latformCheckPairedKeys', but considering other algorithms we may support in the future (eg Curve488) which may have different keys size, it's probably better to implemented in the alg specific function, as you suggested.

javifernandez committed:
Implement a Ed25519 key pair check during the JWK import
https://bugs.webkit.org/show_bug.cgi?id=261323

Reviewed by Žan Doberšek.

This change implemenets the logic to ensure they private and public
keys imported using JWK format are actually compatible, according to
the RFC8032 specification.

There are similar checks for the X25519, already implemented in both
the Mac and Gtk+ ports. The implementation of this check is also present
in the Mac port.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any.worker-expected.txt:
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
* Source/WebCore/crypto/gcrypt/GCryptRFC7748.cpp:
* Source/WebCore/crypto/gcrypt/GCryptRFC8032.cpp: Added.
(WebCore::validateEd25519KeyPair):
* Source/WebCore/crypto/gcrypt/GCryptRFC8032.h: Added.
* Source/WebCore/platform/SourcesGCrypt.txt:

Canonical link: https://commits.webkit.org/267957@main
